src/EoS/Nuclear/CPU_EoS_Nuclear.cpp

```diff
@@ -0,0 +1,579 @@
+{
+
+// check
+#  ifdef GAMER_DEBUG
+#  if ( NCOMP_PASSIVE > 0 )
+   if ( Passive_Code == NULL )   printf( "ERROR : Passive_Code == NULL in %s !!\n", __FUNCTION__ );
+#  endif
+   if ( AuxArray_Flt == NULL )   printf( "ERROR : AuxArray_Flt == NULL in %s !!\n", __FUNCTION__ );
+   if ( AuxArray_Int == NULL )   printf( "ERROR : AuxArray_Int == NULL in %s !!\n", __FUNCTION__ );
+
+   if ( Hydro_CheckNegative(Dens_Code) )
+      printf( "ERROR : invalid input density (%14.7e) at file <%s>, line <%d>, function <%s>\n",
+              Dens_Code, __FILE__, __LINE__, __FUNCTION__ );
+
+   if ( Hydro_CheckNegative(Pres_Code) )
+      printf( "ERROR : invalid input pressure (%14.7e) at file <%s>, line <%d>, function <%s>\n",
+              Pres_Code, __FILE__, __LINE__, __FUNCTION__ );
+#  endif // GAMER_DEBUG
+
+
+
+   const real EnergyShift = AuxArray_Flt[NUC_AUX_ESHIFT   ];
+   const real Dens2CGS    = AuxArray_Flt[NUC_AUX_DENS2CGS ];
+   const real Pres2CGS    = AuxArray_Flt[NUC_AUX_PRES2CGS ];
+   const real sEint2Code  = AuxArray_Flt[NUC_AUX_VSQR2CODE];
+
+   const int  NRho        = AuxArray_Int[NUC_AUX_NRHO  ];
+   const int  NEps        = AuxArray_Int[NUC_AUX_NEPS  ];
+   const int  NYe         = AuxArray_Int[NUC_AUX_NYE   ];
+   const int  NMode       = AuxArray_Int[NUC_AUX_NMODE ];
+
+   int  Mode      = NUC_MODE_PRES;
+   real Dens_CGS  = Dens_Code * Dens2CGS;
+   real Pres_CGS  = Pres_Code * Pres2CGS;
+   real Ye        = Passive_Code[ YE - NCOMP_FLUID ] / Dens_Code;
+   real sEint_CGS = NULL_REAL;
+   real Useless   = NULL_REAL;
+   int  Err       = NULL_INT;
+
+
+// check floating-point overflow and Ye
+#  ifdef GAMER_DEBUG
+   if ( Nuc_Overflow(Dens_CGS) )
+      printf( "ERROR : EoS overflow (Dens_CGS %13.7e, Dens_Code %13.7e, Dens2CGS %13.7e) in %s() !!\n",
+              Dens_CGS, Dens_Code, Dens2CGS, __FUNCTION__ );
+
+   if ( Nuc_Overflow(Pres_CGS) )
+      printf( "ERROR : EoS overflow (Pres_CGS %13.7e, Pres_Code %13.7e, Pres2CGS %13.7e) in %s() !!\n",
+              Pres_CGS, Pres_Code, Pres2CGS, __FUNCTION__ );
+
+   if ( Ye < (real)Table[NUC_TAB_YE][0]  ||  Ye > (real)Table[NUC_TAB_YE][NYe-1] )
+      printf( "ERROR : invalid Ye = %13.7e (min = %13.7e, max = %13.7e) in %s() !!\n",
+              Ye, Table[NUC_TAB_YE][0], Table[NUC_TAB_YE][NYe-1], __FUNCTION__ );
+#  endif // GAMER_DEBUG
+
+
+// invoke the nuclear EoS driver
+   nuc_eos_C_short( Dens_CGS, &sEint_CGS, Ye, &Useless, &Useless, &Pres_CGS, &Useless, &Useless,
+                    EnergyShift, NRho, NEps, NYe, NMode,
+                    Table[NUC_TAB_ALL], Table[NUC_TAB_ALL_MODE], Table[NUC_TAB_RHO], Table[NUC_TAB_EPS],
+                    Table[NUC_TAB_YE], Table[NUC_TAB_TEMP_MODE], Table[NUC_TAB_ENTR_MODE], Table[NUC_TAB_PRES_MODE],
+                    Mode, &Err, NULL_REAL );
+
+// trigger a *hard failure* if the EoS driver fails
+   if ( Err )  sEint_CGS = NAN;
+
+   const real Eint_Code = (  ( sEint_CGS + EnergyShift ) * sEint2Code  ) * Dens_Code;
```

Comment:
Why do we need to add `EnergyShift` here?

Reply:
See above.